File: api/src/main/java/com/sourceplusplus/api/model/QueryTimeFrame.java
package com.sourceplusplus.api.model;

/**
 * Time frames the core supports for artifact metrics/traces.
 * todo: remove class for a more dynamic solution
 *
 * @author <a href="mailto:brandon@srcpl.us">Brandon Fergerson</a>
 * @version 0.3.0
 * @since 0.1.0
 */
public enum QueryTimeFrame {

    LAST_5_MINUTES(5),
    LAST_15_MINUTES(15),
    LAST_30_MINUTES(30),
    LAST_HOUR(60),
    LAST_3_HOURS(60 * 3);

    private final int minutes;

    QueryTimeFrame(int minutes) {
        this.minutes = minutes;
    }

    public int getMinutes() {
        return minutes;
    }

    public static QueryTimeFrame valueOf(int minutes) {
        for (QueryTimeFrame timeFrame : QueryTimeFrame.values()) {
            if (timeFrame.minutes == minutes) {
                return timeFrame;
            }
        }
        throw new IllegalArgumentException("No time frame for minutes: " + minutes);
    }
}

File: protocol/src/commonMain/kotlin/com/sourceplusplus/protocol/portal/QueryTimeFrame.kt
package com.sourceplusplus.protocol.portal

/**
 * todo: description.
 *
 * @since 0.0.1
 * @author [Brandon Fergerson](mailto:bfergerson@apache.org)
 */
enum class QueryTimeFrame(val minutes: Int, val id: String) {
    LAST_5_MINUTES(5, "5_minutes"),
    LAST_15_MINUTES(15, "15_minutes"),
    LAST_30_MINUTES(30, "30_minutes"),
    LAST_HOUR(60, "hour"),
    LAST_3_HOURS(60 * 3, "3_hours"); //todo: id = enum name

    val description = id.toUpperCase().replace("_", " ")

    companion object {
        fun valueOf(minutes: Int): QueryTimeFrame {
            for (timeFrame in values()) {
                if (timeFrame.minutes == minutes) {
                    return timeFrame
                }
            }
            throw IllegalArgumentException("No time frame for minutes: $minutes")
        }
    }
}
